src/ast_editor.ml

```diff
@@ -0,0 +1,585 @@
+        ()
+    in
+    let onDidChangeTextDocument_disposable =
+      Workspace.onDidChangeTextDocument
+        ~listener:(onDidChangeTextDocument_listener instance ~webview ~document)
+        ()
+    in
+    WebviewPanel.onDidDispose webviewPanel
+      ~listener:(fun () ->
+        Ast_editor_state.set_original_mode ast_editor_state true;
+        Disposable.dispose onDidReceiveMessage_disposable;
+        Disposable.dispose onDidChangeTextDocument_disposable)
+      ()
+  in
+  (try transform_to_ast instance ~document ~webview with
+  | User_error err_msg ->
+    send_msg "error" (Jsonoo.Encode.string err_msg |> Jsonoo.t_to_js) ~webview);
+  let options = WebView.options webview in
+  WebviewOptions.set_enableScripts options true;
+  WebView.set_options webview options;
+  let p =
+    let open Promise.Syntax in
+    let+ r = read_html_file () in
+    WebView.set_html webview r
+  in
+  `Promise p
+
+let open_ast_explorer ~uri =
+  let open Promise.Syntax in
+  let+ _ =
+    Vscode.Commands.executeCommand ~command:"vscode.openWith"
+      ~args:
+        [ Uri.t_to_js uri
+        ; Ojs.string_to_js "ast-editor"
+        ; ViewColumn.t_to_js ViewColumn.Beside
+        ]
+  in
+  ()
+
+let replace_document_content ~document ~content =
+  let first_line = TextDocument.lineAt document ~line:0 in
+  let last_line =
+    TextDocument.lineAt document ~line:(TextDocument.lineCount document - 1)
+  in
+  let range =
+    Range.makePositions
+      ~start:(TextLine.range first_line |> Range.start)
+      ~end_:(TextLine.range last_line |> Range.end_)
+  in
+  let edit = WorkspaceEdit.make () in
+  WorkspaceEdit.replace edit
+    ~uri:(TextDocument.uri document)
+    ~range ~newText:content;
+  let (_ : bool Promise.t) = Workspace.applyEdit ~edit in
+  ()
+
+let open_pp_doc instance ~document =
+  let open Promise.Syntax in
+  let ast_editor_state = Extension_instance.ast_editor_state instance in
+  match fetch_pp_code ~document with
+  | Error e -> Promise.return (Error e)
+  | Ok pp_pp_str ->
+    let* doc =
+      Workspace.openTextDocument
+        (`Uri
+          (Uri.parse ("post-ppx: " ^ TextDocument.fileName document ^ "?") ()))
+    in
+    Ast_editor_state.associate_origin_and_pp ast_editor_state
+      ~origin_uri:(TextDocument.uri document)
+      ~pp_doc_uri:(TextDocument.uri doc);
+    replace_document_content ~content:pp_pp_str ~document:doc;
+    let+ (_ : TextEditor.t) =
+      Window.showTextDocument ~document:(`TextDocument doc)
+        ~column:ViewColumn.Beside ()
+    in
+    Ok ()
+
+let reload_pp_doc instance ~document =
+  let open Promise.Syntax in
+  let ast_editor_state = Extension_instance.ast_editor_state instance in
+  let visibleTextEditors = Window.visibleTextEditors () in
+  let origin_uri =
+    match
+      Ast_editor_state.find_original_doc_by_pp_uri ast_editor_state
+        (TextDocument.uri document)
+    with
+    | Some x -> x
+    | None -> failwith "Failed finding the original document URI."
```

Comment:
Again, this exception needs to be something else:

* User_error if the user is responsible or is part of the normal workflow
* Code_error if it's a bug in our extension or we don't know what happened.

Reply:
Again, replaced `failwith` with just an `Error`. 064c5ab58fc14656ef46f80968decdc490acdfae